internal/impl/kafka/franz_writer.go

```diff
@@ -0,0 +1,348 @@
+
+	maxMessageBytesStr, err := conf.FieldString(kfwFieldMaxMessageBytes)
+	if err != nil {
+		return nil, err
+	}
+	maxMessageBytes, err := humanize.ParseBytes(maxMessageBytesStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse max_message_bytes: %w", err)
+	}
+	if maxMessageBytes > uint64(math.MaxInt32) {
+		return nil, fmt.Errorf("invalid max_message_bytes, must not exceed %v", math.MaxInt32)
+	}
+	opts = append(opts, kgo.ProducerBatchMaxBytes(int32(maxMessageBytes)))
+
+	brokerWriteMaxBytesStr, err := conf.FieldString(kfwFieldBrokerWriteMaxBytes)
+	if err != nil {
+		return nil, err
+	}
+	brokerWriteMaxBytes, err := humanize.ParseBytes(brokerWriteMaxBytesStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse broker_write_max_bytes: %w", err)
+	}
+	if brokerWriteMaxBytes > 1<<30 {
+		return nil, fmt.Errorf("invalid broker_write_max_bytes, must not exceed %v", 1<<30)
+	}
+	opts = append(opts, kgo.BrokerMaxWriteBytes(int32(brokerWriteMaxBytes)))
+
+	var compressionPrefs []kgo.CompressionCodec
+	if conf.Contains(kfwFieldCompression) {
+		cStr, err := conf.FieldString(kfwFieldCompression)
+		if err != nil {
+			return nil, err
+		}
+
+		var c kgo.CompressionCodec
+		switch cStr {
+		case "lz4":
+			c = kgo.Lz4Compression()
+		case "gzip":
+			c = kgo.GzipCompression()
+		case "snappy":
+			c = kgo.SnappyCompression()
+		case "zstd":
+			c = kgo.ZstdCompression()
+		case "none":
+			c = kgo.NoCompression()
+		default:
+			return nil, fmt.Errorf("compression codec %v not recognised", cStr)
+		}
+		compressionPrefs = append(compressionPrefs, c)
+	}
+	if len(compressionPrefs) > 0 {
+		opts = append(opts, kgo.ProducerBatchCompression(compressionPrefs...))
+	}
+
+	partitioner := kgo.StickyKeyPartitioner(nil)
+	if conf.Contains(kfwFieldPartitioner) {
+		partStr, err := conf.FieldString(kfwFieldPartitioner)
+		if err != nil {
+			return nil, err
+		}
+		switch partStr {
+		case "murmur2_hash":
+			partitioner = kgo.StickyKeyPartitioner(nil)
+		case "round_robin":
+			partitioner = kgo.RoundRobinPartitioner()
+		case "least_backup":
+			partitioner = kgo.LeastBackupPartitioner()
+		case "manual":
+			partitioner = kgo.ManualPartitioner()
+		default:
+			return nil, fmt.Errorf("unknown partitioner: %v", partStr)
+		}
+	}
+	if partitioner != nil {
+		opts = append(opts, kgo.RecordPartitioner(partitioner))
+	}
+
+	idempotentWrite, err := conf.FieldBool(kfwFieldIdempotentWrite)
+	if err != nil {
+		return nil, err
+	}
+	if !idempotentWrite {
+		opts = append(opts, kgo.DisableIdempotentWrite())
+	}
+
+	timeout, err := conf.FieldDuration(kfwFieldTimeout)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, kgo.ProduceRequestTimeout(timeout))
+
+	return opts, nil
+}
+
+//------------------------------------------------------------------------------
+
+const (
+	rwFieldTopic     = "topic"
+	rwFieldKey       = "key"
+	rwFieldPartition = "partition"
+	rwFieldMetadata  = "metadata"
+	rwFieldTimestamp = "timestamp"
+)
```

Comment:
Should these be prefixed with `fw` (or `kfw` like the ones above) instead of `rw`?